include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -1,44 +1,135 @@
-// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
 // This file is part of the "Nabla Engine".
 // For conditions of distribution and use, see copyright notice in nabla.h
 #ifndef _NBL_BUILTIN_HLSL_BXDF_COMMON_INCLUDED_
 #define _NBL_BUILTIN_HLSL_BXDF_COMMON_INCLUDED_
 
 #include "nbl/builtin/hlsl/limits.hlsl"
 #include "nbl/builtin/hlsl/numbers.hlsl"
-#include "nbl/builtin/hlsl/math/functions.glsl"
+#include "nbl/builtin/hlsl/type_traits.hlsl"
+#include "nbl/builtin/hlsl/concepts.hlsl"
+#include "nbl/builtin/hlsl/ieee754.hlsl"
+#include "nbl/builtin/hlsl/tgmath.hlsl"
+#include "nbl/builtin/hlsl/math/functions.hlsl"
+#include "nbl/builtin/hlsl/cpp_compat/promote.hlsl"
+#include "nbl/builtin/hlsl/bxdf/fresnel.hlsl"
+#include "nbl/builtin/hlsl/vector_utils/vector_traits.hlsl"
 
 namespace nbl
 {
 namespace hlsl
 {
+
 namespace bxdf
 {
 
+// returns unnormalized vector
+template<typename T NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+T computeUnnormalizedMicrofacetNormal(bool _refract, NBL_CONST_REF_ARG(T) V, NBL_CONST_REF_ARG(T) L, typename vector_traits<T>::scalar_type orientedEta)
+{
+    const typename vector_traits<T>::scalar_type etaFactor = (_refract ? orientedEta : 1.0);
+    T tmpH = V + L * etaFactor;
+    tmpH = ieee754::flipSign<T>(tmpH, _refract);
+    return tmpH;
+}
+
+// returns normalized vector, but NaN when result is length 0
+template<typename T NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+T computeMicrofacetNormal(bool _refract, NBL_CONST_REF_ARG(T) V, NBL_CONST_REF_ARG(T) L, typename vector_traits<T>::scalar_type orientedEta)
+{
+    const T H = computeUnnormalizedMicrofacetNormal<T>(_refract,V,L,orientedEta);
+    return hlsl::normalize<T>(H);
+}
+
+// if V and L are on different sides of the surface normal, then their dot product sign bits will differ, hence XOR will yield 1 at last bit
+bool isTransmissionPath(float NdotV, float NdotL)
+{
+    return bool((bit_cast<uint32_t>(NdotV) ^ bit_cast<uint32_t>(NdotL)) & 0x80000000u);
+}
```

Comment:
package up into a `struct compute_microfacet_normal` with method `normalized` and `unnormalized` and static `isTransmissionPath`

let the `create` only create from `V,N,L` and non oriented `eta`